### DISCUSSION TITLE
A question about Jacobian of order parameter wrt. other variables

### DISCUSSION THREAD

**Discussion by bylore** [idaholab/moose#25875] [Q&A Modules: Phase field] [answered]
Hi, I am using Moose for my phase field fracture modelling. There are variables "**ux uy**" and order parameter "**v**"
I used StressDivergenceTensors for Residual **R_u** and Jacobian **J_uu**. and PhaseFieldFractureMechanicsOffDiag for **J_uv**
AllenCahn, ACInterface, TimeDerivative for **R_v** and **J_vv** (I guess).
At this time, everything went well.

And then I built my own class for **J_vu.** When I added this action, the program could not be converged.
Do you have any clues? Does the phase field module has its own way to deal with **J_vu**?

Below is my class code for **J_vu**

```
xxx::ScalarFieldVarCoupleStressOffDiag(
    const InputParameters & parameters)
  : DerivativeMaterialInterface<Kernel>(parameters),
    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
    _ndisp(coupledComponents("displacements")),
    _disp_vars(coupledIndices("displacements")),
    _dF_dc_dstrain(getMaterialProperty<RankTwoTensor>(derivativePropertyNameFirst(_base_name + "stress", _var.name())))
{
}

Real
xxx::computeQpOffDiagJacobian(unsigned int jvar)
{
  if (_ndisp != _mesh.dimension())
    paramError(
        "displacements",
        "The number of variables supplied in 'displacements' must match the mesh dimension.");
  for(unsigned int n=0; n<_ndisp; n++)
  {
    if (jvar == _disp_vars[n])
    {
      Real val = 0.0;
      for (unsigned int k = 0; k < Moose::dim; ++k)
        val += 0.5 * (_dF_dc_dstrain[_qp](n,k) + _dF_dc_dstrain[_qp](k,n)) * _grad_phi[_j][_qp](k);
      return val * _test[_i][_qp];
    }
  }
  
  // Returns if coupled variable is not ux or uy
  return 0.0;
```

**Comment by GiudGiud** [comment]
@laagesen

**Reply by bylore** [reply]
Thank you！

**Comment by laagesen** [comment, marked as answer]
I'd suggest you use this utility to make sure your Jacobian implementation is correct:

https://mooseframework.inl.gov/help/development/analyze_jacobian.html

**Reply by bylore** [reply]
Thank you！
